lib/internal/wasi.js

```diff
@@ -0,0 +1,1301 @@
+      }),
+      path_unlink_file: wrap((fd, pathPtr, pathLen) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_PATH_UNLINK_FILE);
+        if (!stats.path) {
+          return WASI_EINVAL;
+        }
+        this.refreshMemory();
+        const p = Buffer.from(this.memory.buffer, pathPtr, pathLen).toString();
+        fs.unlinkSync(path.resolve(stats.path, p));
+        return WASI_ESUCCESS;
+      }),
+      poll_oneoff: (sin, sout, nsubscriptions, nevents) => {
+        let eventc = 0;
+        let waitEnd = 0;
+        this.refreshMemory();
+        for (let i = 0; i < nsubscriptions; i += 1) {
+          const userdata = this.view.getBigUint64(sin, true);
+          sin += 8;
+          const type = this.view.getUint8(sin);
+          sin += 1;
+          switch (type) {
+            case WASI_EVENTTYPE_CLOCK: {
+              sin += 7; // padding
+              const identifier = this.view.getBigUint64(sin, true);
+              sin += 8;
+              const clockid = this.view.getUint32(sin, true);
+              sin += 4;
+              sin += 4; // padding
+              const timestamp = this.view.getBigUint64(sin, true);
+              sin += 8;
+              const precision = this.view.getBigUint64(sin, true);
+              sin += 8;
+              const subclockflags = this.view.getUint16(sin, true);
+              sin += 2;
+              sin += 6; // padding
+
+              const absolute = subclockflags === 1;
+
+              let e = WASI_ESUCCESS;
+              const n = now(clockid);
+              if (n === null) {
+                e = WASI_EINVAL;
+              } else {
+                const end = absolute ? timestamp : n + timestamp;
+                waitEnd = end > waitEnd ? end : waitEnd;
+              }
+
+              this.view.setBigUint64(sout, userdata, true);
+              sout += 8;
+              this.view.setUint16(sout, e, true); // error
+              sout += 2; // pad offset 2
+              this.view.setUint8(sout, WASI_EVENTTYPE_CLOCK);
+              sout += 1; // pad offset 3
+              sout += 5; // padding to 8
+
+              eventc += 1;
+
+              break;
+            }
+            case WASI_EVENTTYPE_FD_READ:
+            case WASI_EVENTTYPE_FD_WRITE: {
+              sin += 3; // padding
+              const fd = this.view.getUint32(sin, true);
+              sin += 4;
+
+              this.view.setBigUint64(sout, userdata, true);
+              sout += 8;
+              this.view.setUint16(sout, WASI_ENOSYS, true); // error
+              sout += 2; // pad offset 2
+              this.view.setUint8(sout, type);
+              sout += 1; // pad offset 3
+              sout += 5; // padding to 8
+
+              eventc += 1;
+
+              break;
+            }
+            default:
+              return WASI_EINVAL;
+          }
+        }
+
+        this.view.setUint32(nevents, eventc, true);
+
+        while (process.hrtime.bigint() < waitEnd) {
+          // nothing
```

Comment:
Can you add a TODO comment saying to call out to nanosleep/usleep/Sleep?

Reply:
^ I would consider that a blocking comment too – we really shouldn’t be busy looping here…